packages/itwin/tree-widget/src/components/trees/VisibilityTreeEventHandler.ts

```diff
@@ -70,17 +70,27 @@ export class VisibilityTreeEventHandler extends UnifiedSelectionTreeEventHandler
   private _visibilityHandler: IVisibilityHandler | undefined;
   private _selectionPredicate?: VisibilityTreeSelectionPredicate;
   private _listeners = new Array<() => void>();
+  private _isChangingVisibility: boolean;
 
   constructor(params: VisibilityTreeEventHandlerParams) {
     super(params);
     this._visibilityHandler = params.visibilityHandler;
     this._selectionPredicate = params.selectionPredicate;
+    this._isChangingVisibility = false;
 
     if (this._visibilityHandler) {
-      this._listeners.push(this._visibilityHandler.onVisibilityChange.addListener(async (nodeIds, visibilityStatus) => this.updateCheckboxes(nodeIds, visibilityStatus)));
+      this._listeners.push(this._visibilityHandler.onVisibilityChange.addListener(async (nodeIds, visibilityStatus) => {
+        if (this._isChangingVisibility)
+          return;
+        void this.updateCheckboxes(nodeIds, visibilityStatus);
+      }));
     }
 
-    this._listeners.push(this.modelSource.onModelChanged.addListener(async ([_, changes]) => this.updateCheckboxes([...changes.addedNodeIds, ...changes.modifiedNodeIds])));
+    this._listeners.push(this.modelSource.onModelChanged.addListener(async ([_, changes]) => {
+      if (this._isChangingVisibility)
```

Comment:
I think this should not skip updating checkboxes for added nodes. It could lead to newly loaded nodes not having a checkbox visible.